src/convert.jl

```diff
@@ -0,0 +1,228 @@
+    bam_record.n_cigar_op = n_cigar_op
+    bam_record.flags = SAM.flags(record)
+    bam_record.l_seq = SAM.hassequence(record) ? SAM.seqlength(record) : 0
+    bam_record.next_refid = refname_to_refid(get_nextrefname(record), header)
+    bam_record.next_pos = SAM.nextposition(record)-1
+    bam_record.tlen = SAM.templength(record)
+    bam_record
+end
+
+"""
+While all single (i.e., non-array) integer types are stored in SAM as 'i', in BAM any of 'cCsSiI' 
+may be used together with the correspondingly-sized binary integer value, chosen according to the field value's magnitude
+"""
+global const type_to_aux_int_fields = Dict(
+    Int8 => UInt8('c'),
+    UInt8 => UInt8('C'),
+    Int16 => UInt8('s'),
+    UInt16 => UInt8('S'),
+    Int32 => UInt8('i'),
+    UInt32 => UInt8('I'),
+)
+
+function find_smallest_int_type(x)
+    candidates = x < 0 ? (Int8, Int16, Int32) : (UInt8, UInt16, UInt32)
+    idx = findfirst(T -> typemin(T) <= x <= typemax(T) , candidates)
+    isnothing(idx) && error("Coulnd't find Int type for value $(x)")
+    T = candidates[idx]
+
+    T, type_to_aux_int_fields[T]
+end
+
+function encode_aux_field(record::SAM.Record, field::UnitRange{Int64})
+    typ = record.data[first(field)+3]
+    key = record.data[first(field):first(field)+1]
+
+    if typ == UInt8('i')
+        val = record.data[first(field)+5:last(field)] |> String
+        val = parse(Int32, val)
+        T, typ = find_smallest_int_type(val)
+        val = reinterpret(UInt8, [T(val)])
+
+        return vcat(key, typ, val)
+    end
+    if typ == UInt8('Z')
+        val = record.data[first(field)+5:last(field)]
+        return vcat(key, typ, val, 0x00)
+    end
+    if typ == UInt8('A')
+        val = record.data[first(field)+5:last(field)]
+        return vcat(key, typ, val)
+    end
+
+    error("Unsupported tag type $(Char(typ))")
+end
+
+function encode_cigar(record::SAM.Record)
+
+    ops = parse_cigar(record)
+    cigar = Vector{UInt8}(undef, length(ops)*4)
+    k = 1
+    for c in ops
+        cigar_32 = [UInt32(0) + op_to_code[c.op] + (UInt32(c.L) << 4)]
+        for u in reinterpret(UInt8, cigar_32)
+            cigar[k] = u
+            k+=1
+        end
+    end
+    cigar
+end
+
+function encode_sequence(record::SAM.Record)
+    !SAM.hassequence(record) && return UInt8[]
+
+    Nsam = SAM.seqlength(record)
+    Nbam = iseven(Nsam) ? div(Nsam,2) : div(Nsam+1,2)
+
+    # Specs : "When l seq is odd the bottom 4 bits of the last byte are undefined, but we recommend writing these as zero."
+    bamseq = fill(0x00, Nbam)
+    samseq = SAM.sequence(record)
+
+    # SAM
+    # 1 2 3 4 5
+    # A T G G T
+    
+    # BAM
+    # 1  2  3
+    # AT GG T0
+
+    for i in 1:Nbam
+        k = 2i -1
+        val = BioSequences.encode(DNAAlphabet{4}(), samseq[k]) |> UInt8
+        bamseq[i] += val <<4
+        
+        k = 2i
+        k > Nsam && break
+        val = BioSequences.encode(DNAAlphabet{4}(), samseq[k]) |> UInt8
+        bamseq[i] += val
+    end
```

Comment:
BioSequences.encode is an internal symbol. Relying on it is risky.
Also, why not loop over `1:Nsam` instead of `1:Nbam`? Then you know for sure that you don't need to break the loop:
```
for (i, sym) in enumerate(samseq)
    val =  BioSequences.encode(DNAAlphabet{4}(), samseq[k]) |> UInt8
    bamseq[div(i - 1, 2)] |= val << (4 * isodd(i))
end
```

Reply:
Done, with div(i + 1, 2)